[PM-16227] Move import to sdk and enable it in browser/web

quexten opened the pull request:
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-16227

## 📔 Objective

Makes use of the sdk in order to handle import on desktop web and browser. Drops importer code from `desktop_native` since it's not needed there anymore.

## 📸 Screenshots




https://github.com/user-attachments/assets/3e4eb677-5247-4b0c-b36e-5362ef80da53


https://github.com/user-attachments/assets/96906c05-ea68-4856-b4b7-3e1f4d7f060a

Failed import:

https://github.com/user-attachments/assets/2d434548-6223-4758-9bca-76b2b75ea7c6


## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

audreyality commented on libs/angular/src/vault/components/add-edit.component.ts:
```diff
@@ -786,4 +797,68 @@ export class AddEditComponent implements OnInit, OnDestroy {
+              message: this.i18nService.t("sshKeyWrongPassword"),
+            });
+          } else {
+            password = await this.getSshKeyPassword();
+            await this.importSshKeyFromClipboard(password);
```
⚠️ There's no guarantee that the user ever enters the correct password. Using recursion could exhaust the callstack, crashing the consuming code. Consider, instead, running this check in a loop and exiting if the number of failed attempts exceeds a threshold.

💭 This is the kind of code that rxjs is quite good at managing. The solution is likely non-trivial, however, as you'll need to materialize and dematerialize the stream to properly implement the retry logic. I gave it a try, but didn't feel comfortable providing the code for use in logic that isn't unit tested. 🫠

quexten replied:
The desktop-only implementation had a similar bug where a user would be indefinitely prompted. I do think it's unlikely that a user would re-enter the password enough times to actually bear the risk of exhausting the callstack, but I agree it's more readable in a loop.

audreyality commented on libs/vault/src/cipher-form/components/sshkey-section/sshkey-section.component.ts:
```diff
@@ -79,4 +90,71 @@ export class SshKeySectionComponent implements OnInit {
     });
   }
+
+  async importSshKeyFromClipboard(password: string = "") {
```
♻️ This function is replicated with minor alterations across several components. Consider extracting the decision making into one or more shared functions.

quexten replied:
To be honest, _after rewriting it_, it is so much shorter that moving this out to a service, just for the two consumers, seems like it would lead to less clarity of the code.

audreyality reviewed (changes requested):
Requested changes marked with ⚠️.  
Review-blocking concerns that may not require changes marked with 🤔.
Everything else is commentary.

quexten commented on libs/angular/src/vault/components/add-edit.component.ts:
```diff
@@ -786,4 +797,68 @@ export class AddEditComponent implements OnInit, OnDestroy {
+              variant: "error",
+              title: "",
+              message: this.i18nService.t("sshKeyWrongPassword"),
+            });
```
Agreed, should be in a better state now.

quexten commented on libs/angular/src/vault/components/add-edit.component.ts:
```diff
@@ -786,4 +797,68 @@ export class AddEditComponent implements OnInit, OnDestroy {
+        title: "",
+        message: this.i18nService.t("sshKeyPasted"),
+      });
+    } catch (e) {
```
Agreed, and updated. Should be more readable now.

audreyality commented on libs/common/src/models/export/ssh-key.export.ts:
```diff
@@ -24,9 +25,10 @@ export class SshKeyExport {
-    domain.privateKey = req.privateKey != null ? new EncString(req.privateKey) : null;
-    domain.publicKey = req.publicKey != null ? new EncString(req.publicKey) : null;
-    domain.keyFingerprint = req.keyFingerprint != null ? new EncString(req.keyFingerprint) : null;
+    const parsedKey = import_ssh_key(req.privateKey);
```
@djsmith85 - I think this is a place where we'll need your importer expertise.

djsmith85 replied:
The current behaviour is that, if any error on any item occurs, the whole import is aborted. We might want to change that in the future, if we can provide a detailed import report, but for now it just aborts, to not cause partial imports without a notice to the user.

@quexten Please ensure during dev/testing that an invalid ssh key will throw and show which import-item is causing the issue. If you are also able to provide a detailed/informative error messages, users will appreciate it.

quexten replied:
Ok, the error message from the import is now directly exposed to the user. It could use clarification though, since it is unclear which item this is referring to, and the error message doesn't actually say that it is an *ssh key*. The latter part can be changed in the sdk, for the former part. Is that blocking for this PR?

audreyality commented on libs/angular/src/vault/components/add-edit.component.ts:
```diff
@@ -786,4 +799,69 @@ export class AddEditComponent implements OnInit, OnDestroy {
+            title: "",
+            message: this.i18nService.t(this.sshImportErrorVariantToI18nKey(error.variant)),
+          });
+          return;
+        }
+        continue;
+      }
+      break;
```
⛏️ Mixing 3 different control flow keywords here leads me to believe there's a better way to state this.

quexten replied:
Splitting the first attempt out seemed to help simplify this a bit. It's longer now but more readable.

audreyality replied:
Yes! The new code is much more clear! Thank you! 😄

audreyality commented on libs/angular/src/vault/components/add-edit.component.ts:
```diff
@@ -786,4 +797,68 @@ export class AddEditComponent implements OnInit, OnDestroy {
+          } else {
+            password = await this.getSshKeyPassword();
+            await this.importSshKeyFromClipboard(password);
+          }
```
This doesn't need to be solved now--♻️ identifies tech debt. If you're interested in it, though, you might consider embedding an enum in the `Ok` block of the result.

```
enum SshImport {
  Imported(value),
  MissingPassword, // expected failure case
  InvalidPassword, // expected failure case
}

fn method(...) : Result<SshImport, Exception> {
  // ...
}
```

quexten replied:
Wasm-bindgen does not support enums with values at the moment (otherwise Result would also be trivial to represent), but only C-style enums: https://github.com/rustwasm/wasm-bindgen/issues/2407. So for passing to typescript, this would need to be converted to some struct, similar to what was done in the desktop-only implementation.

audreyality replied:
Or you could use [Tsify-next](https://github.com/siefkenj/tsify?tab=readme-ov-file#enum), which is already included as a dependency.

quexten commented on apps/browser/src/_locales/en/messages.json:
```diff
@@ -4903,5 +4903,32 @@
+  "importSshKeyFromClipboard": {
+    "message": "Import key from clipboard"
+  },
+  "sshKeyPasted": {
```
Fixed

quexten commented on apps/web/src/locales/en/messages.json:
```diff
@@ -10201,6 +10200,36 @@
+  "importSshKeyFromClipboard": {
+    "message": "Import key from clipboard"
+  },
+  "sshKeyPasted": {
```
Fixed

quexten commented on libs/vault/src/cipher-form/components/sshkey-section/sshkey-section.component.ts:
```diff
@@ -100,6 +114,73 @@ export class SshKeySectionComponent implements OnInit {
+    const key = await this.platformUtilsService.readFromClipboard();
+
+    let parsedKey: SshKey = null;
+    for (let attempt = 0; attempt <= ATTEMPTS; attempt++) {
```
Yeah, makes sense. Seems eslint forbids `while(true) though, and requires `for (;;) {` instead..

quexten commented on libs/importer/src/importers/bitwarden/bitwarden-json-importer.ts:
```diff
@@ -40,11 +41,19 @@ export class BitwardenJsonImporter extends BaseImporter implements Importer {
   }
 
   async parse(data: string): Promise<ImportResult> {
+    // parsing ssh keys needs an sdk initialized
+    // remove this once the sdk gets initialized automatically
+    const isRunningInJest = typeof jest !== "undefined";
+    if (!isRunningInJest) {
+      await firstValueFrom(this.sdkService.client$);
+    }
```
Fixed.

mzieniukbw commented on libs/vault/src/cipher-form/components/sshkey-section/sshkey-section.component.ts:
```diff
@@ -119,32 +119,60 @@ export class SshKeySectionComponent implements OnInit {
-      break;
+    }
+
+    if (parsedKey == null) {
```
i don't think we can ever reach this. The conditions above either return or break the looping with valid (non null or undefined) `parsedKey`

quexten replied:
I agree, nice catch. Removed.

Patrick-Pimentel-Bitwarden commented:
I suspect the auth team got added to this by a stale code ownership file as I'm not seeing which files falls under our domain nor is github showing that we are a code owner for this pr. Removing myself from the review, if you believe that is in error please add me back.

gbubemismith commented on libs/vault/src/cipher-form/components/sshkey-section/sshkey-section.component.ts:
```diff
@@ -100,6 +114,92 @@ export class SshKeySectionComponent implements OnInit {
   }
 
+  private async importUsingSdk(key: string, password: string): Promise<SshKey> {
+    await firstValueFrom(this.sdkService.client$);
+    return import_ssh_key(key, password);
```
❓ I noticed this code uses `await firstValueFrom(this.sdkService.client$);` without using the client value. Is there a reason we're not piping through the client like other SDK usages?

gbubemismith replied:
Oh thank you. I Just started working on the SDK 👍🏼

gbubemismith commented on libs/vault/src/cipher-form/components/sshkey-section/sshkey-section.component.ts:
```diff
@@ -100,6 +114,92 @@ export class SshKeySectionComponent implements OnInit {
+
+    let isPasswordProtectedSshKey = false;
+
+    let parsedKey: SshKey = null;
```
⛏️ (non-blocking) small change for when we enable strict TS on this file
```suggestion
    let parsedKey: SshKey | null = null;
```

gbubemismith commented on libs/angular/src/vault/components/add-edit.component.ts:
```diff
@@ -812,6 +826,90 @@ export class AddEditComponent implements OnInit, OnDestroy {
+              message: this.i18nService.t(this.sshImportErrorVariantToI18nKey(error.variant)),
+            });
+            return;
+          }
+        }
+      }
+    }
+
+    this.cipher.sshKey.privateKey = parsedKey.private_key;
+    this.cipher.sshKey.publicKey = parsedKey.public_key;
+    this.cipher.sshKey.keyFingerprint = parsedKey.key_fingerprint;
+    this.toastService.showToast({
+      variant: "success",
+      title: "",
+      message: this.i18nService.t("sshKeyImported"),
+    });
+  }
+
+  private sshImportErrorVariantToI18nKey(variant: string): string {
+    switch (variant) {
+      case "ParsingError":
+        return "invalidSshKey";
+      case "UnsupportedKeyType":
+        return "sshKeyTypeUnsupported";
+      case "PasswordRequired":
+      case "WrongPassword":
+        return "sshKeyWrongPassword";
+      default:
+        return "errorOccurred";
+    }
+  }
+
+  async getSshKeyPassword(): Promise<string> {
+    const dialog = this.dialogService.open<string>(SshKeyPasswordPromptComponent, {
+      ariaModal: true,
+    });
+
+    return await lastValueFrom(dialog.closed);
+  }
+
```
🎨 Looks like we are repeating complex logic in different components, is it possible to have this in a service so it can be reusable?

quexten replied:
Makes sense, will update this tomorrow.

quexten replied:
Moved to a service, and added tests covering the basic flows of the service.

gbubemismith reviewed (commented):
I have left a few suggestions and one question on how the SDK client is used

quexten commented on libs/vault/src/services/default-ssh-import-prompt.service.ts:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Used to import ssh keys and prompt for their password.
+ */
+export class DefaultSshImportPromptService {
```
Fixed!

audreyality commented on libs/importer/src/importers/bitwarden/bitwarden-json-importer.ts:
```diff
@@ -47,12 +47,7 @@ export class BitwardenJsonImporter extends BaseImporter implements Importer {
-    if (!isRunningInJest) {
-      await firstValueFrom(this.sdkService.client$);
-    }
+    await firstValueFrom(this.sdkService.client$);
```
⛏️ Seems like you went too far with this change. The comment should still be included, correct? If so, make sure you call out the tech debt with a FIXME.

```suggestion
    // initializes the SDK
    // FIXME: remove this once the SDK gets initialized automatically
    await firstValueFrom(this.sdkService.client$);
```

🤔 If this line is no longer needed for initialization, what does it do? Why isn't the value returned by `client$` used?

quexten replied:
> 🤔 If this line is no longer needed for initialization, what does it do? Why isn't the value returned by client$ used?

Yeah, I agree this comment makes sense. It is unfortunate that this needs to be done using such a confusing line of code at the moment.

audreyality reviewed (commented):
👍🏻 The changes are looking good! I'm glad that the jest detection logic was removed from the importer. Just got one nit to pick.

quexten commented:
Brought up-to-date with main. Seems the macos build failures are unrelated and happening on all PRs.

quexten commented:
@Patrick-Pimentel-Bitwarden 

> I suspect the auth team got added to this by a stale code ownership file as I'm not seeing which files falls under our domain nor is github showing that we are a code owner for this pr. Removing myself from the review, if you believe that is in error please add me back.

Not sure what happened with the codeowners at this point, but after addressing feedback from other reviewers, auth is once again in the codeowers for this changes for `apps/web/src/app/auth/settings/emergency-access/view/emergency-add-edit-cipher.component.ts`, so I've added you back as reviewer.

mzieniukbw commented on libs/vault/src/services/default-ssh-import-prompt.service.ts:
```diff
@@ -0,0 +1,109 @@
+      const error = e as SshKeyImportError;
+      if (error.variant === "PasswordRequired" || error.variant === "WrongPassword") {
+        isPasswordProtectedSshKey = true;
+      } else {
```
🎨 could you add a test case coverage for this ?

quexten replied:
Is this not covered here? https://github.com/bitwarden/clients/blob/de0d9646d8e8d8e7d39e79954f494e1f74c111b5/libs/vault/src/services/ssh-import-prompt.service.spec.ts#L58-L70

quexten commented on libs/vault/src/services/default-ssh-import-prompt.service.ts:
```diff
@@ -0,0 +1,109 @@
+          break;
+        } catch (e) {
+          const error = e as SshKeyImportError;
+          if (error.variant !== "WrongPassword") {
```
Added tests for the error cases

Patrick-Pimentel-Bitwarden reviewed (approved):
👍  Auth changes look good as well as the rest of the pr.

quexten commented:
Resolved merge conflicts

Patrick-Pimentel-Bitwarden reviewed (dismissed):
👍 Great work!

coltonhurst reviewed (approved):
Autofill removals look good!
